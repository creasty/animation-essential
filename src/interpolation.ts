/**
 * Clamps a value to be within a given range.
 *
 * @param {number} v - The value to clamp.
 * @param {number} min - The minimum value in the range.
 * @param {number} max - The maximum value in the range.
 * @returns {number} The clamped value.
 */
export function clamp(v: number, min: number, max: number): number {
  if (v < min) return min;
  if (v > max) return max;
  return v;
}

/**
 * Linearly interpolates between two values.
 *
 * @param {number} a - The first value.
 * @param {number} b - The second value.
 * @param {number} t - The interpolation factor between 0 and 1.
 * @returns {number} The interpolated value.
 */
export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

/**
 * Calculates a smooth step function with an easing effect.
 *
 * @param {number} a - The lower bound of the smooth step function.
 * @param {number} b - The upper bound of the smooth step function.
 * @param {number} x - The input value to the smooth step function.
 * @returns {number} The smoothed output value between 0 and 1.
 */
export function smoothStep(a: number, b: number, x: number): number {
  const t = clamp((x - a) / (b - a), 0, 1);
  return t * t * (3 - 2 * t);
}

/**
 * Linearly interpolates between two consecutive values in the points given in array.
 *
 * @param {number[]} points - The array of points. Must have at least two elements.
 * @param {number} t - The interpolation factor between 0 and [the number of points] - 1.
 * @returns {number} The interpolated value.
 */
export function multiPointLerp(points: number[], t: number): number {
  const l = points.length;
  if (l < 2) throw new TypeError("multiPointLerp requires at least two points");

  t = clamp(t, 0, l - 1);
  const intPart = Math.floor(t);
  const fracPart = t - intPart;
  return lerp(points[intPart], points[Math.min(intPart + 1, l - 1)], fracPart);
}

/**
 * Combines the integer part of an input value with the result of applying a mix function to its fractional part.
 * It comes in very handy when used with {@link multiPointLerp}.
 *
 * @param {number} v - The input value.
 * @param {function} mix - The mix function.
 * @returns {number} The mixed value.
 */
export function fractionalMix(v: number, mix: (t: number) => number): number {
  const intPart = Math.floor(v);
  const fracPart = v - intPart;
  return intPart + mix(fracPart);
}
