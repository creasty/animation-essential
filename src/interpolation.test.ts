import { clamp, lerp, smoothStep, fractionalMix, multiPointLerp } from "./interpolation";

function generateArray(start: number, end: number, step: number) {
  const result = [];
  for (let i = start; i <= end; i += step) {
    result.push(i);
  }
  return result;
}

describe("clamp", () => {
  test("clamps a value between a min and max", () => {
    expect(clamp(-1000, 1, 9)).toBe(1);
    expect(clamp(0, 1, 9)).toBe(1);
    expect(clamp(1, 1, 9)).toBe(1);
    expect(clamp(2, 1, 9)).toBe(2);
    expect(clamp(8, 1, 9)).toBe(8);
    expect(clamp(9, 1, 9)).toBe(9);
    expect(clamp(1000, 1, 9)).toBe(9);
  });
});

describe("lerp", () => {
  test("interpolates a value between two points", () => {
    expect(lerp(1, 9, 0)).toBe(1);
    expect(lerp(1, 9, 0.25)).toBe(3);
    expect(lerp(1, 9, 0.5)).toBe(5);
    expect(lerp(1, 9, 0.75)).toBe(7);
    expect(lerp(1, 9, 1)).toBe(9);
  });
});

describe("smoothStep", () => {
  test.each([
    [0, 1, -1, 0],
    [0, 1, 0, 0],
    [0, 1, 0.5, 0.5],
    [0, 1, 1, 1],
    [0, 1, 2, 1],
    [-1, 1, -2, 0],
    [-1, 1, -1, 0],
    [-1, 1, 0, 0.5],
    [-1, 1, 1, 1],
    [-1, 1, 2, 1],
    [2, 4, 1, 0],
    [2, 4, 2, 0],
    [2, 4, 3, 0.5],
    [2, 4, 4, 1],
    [2, 4, 5, 1],
  ])("smoothStep(%f, %f, %f) returns %f", (a, b, x, expected) => {
    expect(smoothStep(a, b, x)).toBeCloseTo(expected);
  });
});

describe("multiPointLerp", () => {
  test("requires at least two points", () => {
    expect(() => {
      multiPointLerp([], 0);
    }).toThrowError(TypeError);
    expect(() => {
      multiPointLerp([0], 0);
    }).toThrowError(TypeError);
    expect(() => {
      multiPointLerp([0, 1], 0);
    }).not.toThrowError(TypeError);
  });

  test("interpolates between two values", () => {
    const ts = generateArray(0, 1, 0.1);
    const points = [1, 9];
    expect(ts.map((t) => multiPointLerp(points, t))).toMatchSnapshot();
  });

  test("interpolates between two consecutive values in multiple points", () => {
    const ts = generateArray(0, 3, 0.1);
    const points = [1, 20, 300, 4000];
    expect(ts.map((t) => multiPointLerp(points, t))).toMatchSnapshot();
  });
});

describe("fractionalMix", () => {
  test("applies the mix function only to the fractional part of input", () => {
    const values = generateArray(0, 3, 0.1);
    const mixFunc = (t: number) => t * t;
    expect(values.map((v) => fractionalMix(v, mixFunc))).toMatchSnapshot();
  });
});
