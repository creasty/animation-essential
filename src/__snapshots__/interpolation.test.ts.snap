// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fractionalMix applies the mix function only to the fractional part of input 1`] = `
[
  0,
  0.010000000000000002,
  0.04000000000000001,
  0.09000000000000002,
  0.16000000000000003,
  0.25,
  0.36,
  0.48999999999999994,
  0.6399999999999999,
  0.8099999999999998,
  0.9999999999999998,
  1.01,
  1.04,
  1.09,
  1.1600000000000001,
  1.2500000000000002,
  1.3600000000000003,
  1.4900000000000007,
  1.6400000000000008,
  1.810000000000001,
  2,
  2.0100000000000002,
  2.04,
  2.0900000000000003,
  2.1600000000000006,
  2.250000000000001,
  2.360000000000001,
  2.4900000000000015,
  2.640000000000002,
  2.8100000000000023,
]
`;

exports[`multiPointLerp interpolates between two consecutive values in multiple points 1`] = `
[
  1,
  2.9000000000000004,
  4.800000000000001,
  6.700000000000001,
  8.600000000000001,
  10.5,
  12.4,
  14.299999999999999,
  16.2,
  18.099999999999998,
  19.999999999999996,
  47.999999999999964,
  75.99999999999999,
  104.00000000000001,
  132.00000000000006,
  160.00000000000006,
  188.00000000000009,
  216.0000000000001,
  244.00000000000014,
  272.00000000000017,
  300.00000000000165,
  670.000000000002,
  1040.0000000000023,
  1410.0000000000027,
  1780.000000000003,
  2150.000000000003,
  2520.0000000000036,
  2890.000000000004,
  3260.000000000004,
  3630.0000000000045,
]
`;

exports[`multiPointLerp interpolates between two values 1`] = `
[
  1,
  1.8,
  2.6,
  3.4000000000000004,
  4.2,
  5,
  5.8,
  6.6,
  7.3999999999999995,
  8.2,
  9,
]
`;
